curl: run tests in passthru

Artturin opened the pull request:
curl is built many times during a stdenv rebuild

they are a bottleneck

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin committed:
curl: use finalAttrs

allows using the correct package in passthru

remove unneeded replaceStrings in urls

Artturin committed:
phpPackage.mkExtension: makeOverridable

Artturin committed:
curl: put tests in tests.withCheck

curl is built many times during a stdenv rebuild

they are a bottleneck

roberth commented on pkgs/tools/networking/curl/default.nix:
```diff
@@ -177,16 +180,22 @@ stdenv.mkDerivation rec {
       # Additional checking with support http3 protocol.
-      inherit (nixosTests) nginx-http3;
+      # nginx-http3 = useThisCurl nixosTests.nginx-http3;
+      nginx-http3 = nixosTests.nginx-http3;
```
In the future (#176557), `nixosTests.nginx-http3.extend { defaults = { /* NixOS module */ }; }`

(fyi)

roberth commented on pkgs/tools/networking/curl/default.nix:
```diff
@@ -177,16 +180,22 @@ stdenv.mkDerivation rec {
-      pycurl = python3.pkgs.pycurl;
+      curlWithTests = finalAttrs.finalPackage.overrideAttrs (_: { doCheck = true; });
+      curlpp = useThisCurl curlpp;
+      coeurl = useThisCurl coeurl;
```
```nix
      fetchpatch = tests.fetchpatch.simple.override { fetchpatch = fetchpatch.override { fetchurl = useThisCurl fetchurl; }; };
```

where `tests` is `pkgs.tests`.

Haven't tried this, but would be pretty great to have something like this.
Also would be nice to have a direct fetchurl test. :shrug:

roberth replied:
I didn't actually want to repeat `.simple` here, but I don't think we have an `overridePackages` to complement `callPackages` yet. That could use `recurseForDerivations` to do the right thing.

roberth commented on pkgs/tools/networking/curl/default.nix:
```diff
@@ -177,16 +180,22 @@ stdenv.mkDerivation rec {
-      ocaml-curly = ocamlPackages.curly;
-      php-curl = phpExtensions.curl;
-      pycurl = python3.pkgs.pycurl;
+      curlWithTests = finalAttrs.finalPackage.overrideAttrs (_: { doCheck = true; });
```
```suggestion
      withCheck = finalAttrs.finalPackage.overrideAttrs (_: { doCheck = true; });
```

So two things
 - `curl` is a bit redundant. `withTests` might be more meaningful to other code.
 - `tests` -> `check`, as we're setting `doCheck`

roberth reviewed (commented):
Awesome!

SuperSandro2000 commented:
ofborg is failing for some reason but I cannot look it up because the log is not loading.

Mindavi commented:
I knew someone would come in and change running the tests again.

See https://github.com/NixOS/nixpkgs/pull/170828 for some previous discussion and why they were enabled.

roberth commented on pkgs/tools/networking/curl/default.nix:
```diff
@@ -177,16 +182,23 @@ stdenv.mkDerivation rec {
-      ocaml-curly = ocamlPackages.curly;
-      php-curl = phpExtensions.curl;
-      pycurl = python3.pkgs.pycurl;
+      withCheck = finalAttrs.finalPackage.overrideAttrs (_: { doCheck = true; });
```
This one should be a channel blocker if it fails. It should be added to release.nix for the nixpkgs and nixos jobs.

Artturin replied:
not sure i did it correctly

Artturin replied:
```
error: attribute 'tests' missing

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-4/pkgs/top-level/release.nix:100:15:

           99|               jobs.stdenv.x86_64-linux
          100|               jobs.curl.tests.withCheck.x86_64-linux
             |               ^
          101|               jobs.cargo.x86_64-linux
```

are passthru attributes stripped?

roberth replied:
yeah release.nix is kind of a mess (or maybe I'm unfamiliar) and it strips all packages down using `hydraJobs`. It's supposed to help that evaluator release memory.
If you add the passthru tests as "normal" packages somewhere in `pkgs`, you won't have to deal with the release.nix code as much.

Maybe do something like this in `pkgs/test/default.nix`?

```nix

corePackageTests = recurseIntoAttrs {
  curl = recurseIntoAttrs pkgs.curl.tests;
};

```

That should give you `pkgs.tests.corePackageTests`, which you can use in the release files.

Artturin committed:
release: add tests.packageTestsForChannelBlockers.curl.withCheck as a channel blocker

we can't add 'nixpkgs.curl.tests' to hydra jobs due to 'tests' (and 'passthru') being stripped

TODO: add a function in lib-release.nix to get derivations and add `.x86_64-linux` to them
then we can just point release files to nixpkgs.tests.packageTestsForChannelBlockers instead of
nixpkgs.tests.packageTestsForChannelBlockers.curl.withCheck

Artturin commented on pkgs/tools/networking/curl/default.nix:
```diff
@@ -177,16 +182,23 @@ stdenv.mkDerivation rec {
 
-  passthru = {
+  passthru = let
+    useThisCurl = attr: attr.override { curl = finalAttrs.finalPackage; };
```
@risicle https://github.com/NixOS/nixpkgs/pull/170828#issuecomment-1113568253
> The biggest problem I see with this though, is that fails to properly deal with a package's numerous variants because of the self-referential problems passthru.tests

with `finalAttrs.finalPackage` that issue has been solved

risicle replied:
How has this been around for over a year and I've only just learned of it??

Artturin replied:
I merged the pr 1 month ago https://github.com/NixOS/nixpkgs/pull/119942

Izorkin commented on pkgs/tools/networking/curl/default.nix:
```diff
@@ -177,16 +182,23 @@ stdenv.mkDerivation rec {
       # Additional checking with support http3 protocol.
-      inherit (nixosTests) nginx-http3;
+      # nginx-http3 = useThisCurl nixosTests.nginx-http3;
+      nginx-http3 = nixosTests.nginx-http3;
```
The test runs successfully on my system. Checking on 458d91e branch.

Artturin replied:
It fails on 2 machines that I tried on and it fails on ofborg


```
client #   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
client #                                  Dload  Upload   Total   Spent    Left  Speed
client #   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 192.168.2.101:443...
client # * Connect socket 5 over QUIC to 192.168.2.101:443
server # [    5.526163] dhcpcd[679]: eth0: soliciting an IPv6 router
client # * ngtcp2: recvfrom() unexpectedly returned -1
client # * connect to 192.168.2.101 port 443 failed: Failure when receiving data from the peer
client # * Failed to connect to acme.test port 443 after 174 ms: Failure when receiving data from the peer
client #   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
client # * Closing connection 0
client # curl: (56) ngtcp2: recvfrom() unexpectedly returned -1
```

Izorkin replied:
This error fixed in this PR - https://github.com/NixOS/nixpkgs/pull/179055/files#diff-36ed05d3c94024e5ae75789f5251f068d8da25c1a60f2e83adc43e80835a9a7d

Artturin replied:
i pulled the fix here for now, will drop before merge

roberth commented on pkgs/test/default.nix:
```diff
@@ -24,6 +24,11 @@ with pkgs;
 
   config = callPackage ./config.nix { };
 
+  # we can't add 'nixpkgs.curl.tests' to hydra jobs due to 'passthru' being stripped
+  passthruTestsForChannelBlockers = recurseIntoAttrs {
```
```suggestion
  # we can't add 'nixpkgs.curl.tests' to hydra jobs due to 'tests' (and 'passthru') being stripped
  packageTestsForChannelBlockers = recurseIntoAttrs {
```

The fact that they're on a package is more significant than `passthru`, which is just an implementation detail of `mkDerivation`.

`passthru` was introduced in https://github.com/NixOS/nixpkgs/commit/1014ca2ad55761b54366d853da45a11194d4f531 (2006) and only exposed the attrs inside of it. The behavior `passthru.passthru = passthru` was only added later for inspection https://github.com/NixOS/nixpkgs/commit/84fba68006259f1732a9580ba2bdcbc300300c89:

> to distinguish after the fact which attributes were part of the derivation and which came from passthru.

However, the root cause of any such ambiguity comes from the mistake that `mkDerivation` exposes all its argument attributes in the returned package attrset.

Since RFC 92, not all packages will be defined through `mkDerivation` anymore, so we should try not to perpetuate its jargon and keep things simple.

SuperSandro2000 commented:
@ofborg eval

Artturin commented:
the hydra jobs for `tests.packageTestsForChannelBlockers` look like this

```
$ ./maintainers/scripts/rebuild-amount.sh  --print HEAD~1
Estimating rebuild amount by counting changed Hydra jobs (parallel=unset).
      5 x86_64-darwin
     11 x86_64-linux

nixos-install-tools.x86_64-linux                                                      /nix/store/mddaircl7p6ka6v30sl3hnbkp30xp22n-nixos-install-tools-22.11pre-git
tests.nixos-functions.nixos-test.x86_64-linux                                         /nix/store/g5wq3p43akl3c0zszp676q4nq5nyh5vv-nixos-system-nixos-test
tests.packageTestsForChannelBlockers.curl.coeurl.x86_64-darwin                        /nix/store/x3x1mq7bfd6rb8hi2r84g5n1mfafpf4h-coeurl-0.2.0
tests.packageTestsForChannelBlockers.curl.coeurl.x86_64-linux                         /nix/store/ja1wpybr47p53il1hrqnajmiisrg56si-coeurl-0.2.0
tests.packageTestsForChannelBlockers.curl.curlpp.x86_64-linux                         /nix/store/mlqrdavazq0pkfvaxz9ksr540pfy7bdb-curlpp-0.8.1
tests.packageTestsForChannelBlockers.curl.fetchpatch.x86_64-linux                     /nix/store/gh62y23d9xjx38xn4mkwfmy6gmpplyw2-source-salted-z09f6i1ny721
tests.packageTestsForChannelBlockers.curl.haskell-curl.x86_64-darwin                  doc=/nix/store/iv8lyrn13x472gxfiv4bbfx2fp007bmi-curl-1.3.8-doc;/nix/store/k2q9dm1sd7pc16ydn3xkn02h700vmkpd-curl-1.3.8
tests.packageTestsForChannelBlockers.curl.haskell-curl.x86_64-linux                   doc=/nix/store/i26k3nsqm5i4w29ywizngkmnyvd9rydd-curl-1.3.8-doc;/nix/store/8id3zbivfiiy856w24675v2i7vrsa7v7-curl-1.3.8
tests.packageTestsForChannelBlockers.curl.nginx-http3.x86_64-linux                    /nix/store/isd4bnd78js04zm4mhhxlbfbdlczf01j-vm-test-run-nginx-http3
tests.packageTestsForChannelBlockers.curl.ocaml-curly.x86_64-darwin                   /nix/store/vs7zsly4b9ynkmky48hipbvrmdamxnxa-ocaml4.13.1-curly-0.2.0
tests.packageTestsForChannelBlockers.curl.ocaml-curly.x86_64-linux                    /nix/store/pr8128y509li57qqzf6misk0a7h18ci4-ocaml4.13.1-curly-0.2.0
tests.packageTestsForChannelBlockers.curl.php-curl.x86_64-linux                       dev=/nix/store/f7pa423hkdpzqvii0xdbk7s3r0ljqczw-php-curl-8.1.7-dev;/nix/store/aijj4f0f6iskri7j0i955f6rvg7f43ls-php-curl-8.1.7
tests.packageTestsForChannelBlockers.curl.pycurl.x86_64-darwin                        /nix/store/8rim82j9akdcq55jrnaj9ggwsz6400bg-python3.10-pycurl-7.45.1
tests.packageTestsForChannelBlockers.curl.pycurl.x86_64-linux                         /nix/store/dyr6m5ac757jvgpvj72w4k6fi3i6r9n6-python3.10-pycurl-7.45.1
tests.packageTestsForChannelBlockers.curl.withCheck.x86_64-darwin                     bin=/nix/store/bc5z41yg99n8f58hjfvrcpp552n872qs-curl-7.83.1-bin;dev=/nix/store/ly0q61iaf2gfssf9aab4mqb3qrgbkil1-curl-7.83.1-dev;devdoc=/nix/store/wik2qchahxksfmdcrgamc5pmdpzhz811-curl-7.83.1-devdoc;man=/nix/store/ccrjr4z4p7l9b5c2h465xbyiwna80w6h-curl-7.83.1-man;/nix/store/81axcdapacnw0dxjrw0f00brqh2x3b9c-curl-7.83.1
tests.packageTestsForChannelBlockers.curl.withCheck.x86_64-linux                      bin=/nix/store/8w50x3ygiymlxmy36gj7y6xsg9kqvnjv-curl-7.83.1-bin;debug=/nix/store/zpf9j93lc3br13wg9adh2la9ksay0npz-curl-7.83.1-debug;dev=/nix/store/zv467z7cq9yl23igax527z64krp4vr4r-curl-7.83.1-dev;devdoc=/nix/store/qfa4ifwbm6sq11qz1q2b924nb986k2q0-curl-7.83.1-devdoc;man=/nix/store/03kkz40l7yc4qgdaf3kp4b86z6zjzzx3-curl-7.83.1-man;/nix/store/qz9lczmn7h5l9gz7yw57kjl1q7msy275-curl-7.83.1
```

vcunat commented:
Error: https://hydra.nixos.org/jobset/nixos/staging-small#tabs-errors

vcunat commented:
BTW, this is in `staging-next` now, PR #181133

Artturin commented:
reverted the release tests in https://github.com/NixOS/nixpkgs/pull/181214
